testng-core/src/test/kotlin/test/hook/HookableTest.kt

```diff
@@ -0,0 +1,285 @@
+    )
+
+    @Test
+    fun configurableFailure() {
+        val listener = TestResultsCollector()
+        create(ConfigurableFailureSample::class.java).run {
+            addListener(listener)
+            run()
+        }
+        assertThat(listener.getPassedConfigNames()).isEmpty()
+        assertThat(listener.getFailedConfigs()).hasSize(1)
+        val failedConfigResult = listener.getFailedConfigs()[0]
+        assertThat(failedConfigResult.getAttribute(HOOK_INVOKED_ATTRIBUTE)).isNotNull
+        assertThat(failedConfigResult.getAttribute(HOOK_METHOD_INVOKED_ATTRIBUTE)).isNull()
+        assertThat(listener.getSkippedConfigs()).hasSize(3)
+        val assertions = SoftAssertions()
+        listener.getSkippedConfigs()
+            .forEach { each ->
+                assertions.assertThat(each.getAttribute(HOOK_INVOKED_ATTRIBUTE)).isNull()
+                assertions.assertThat(each.getAttribute(HOOK_METHOD_INVOKED_ATTRIBUTE)).isNull()
+            }
+        assertions.assertAll()
+    }
+
+    @Test
+    fun configurableFailureWithStatusAltered() {
+        val listener = TestResultsCollector()
+        create(ConfigurableFailureWithStatusAlteredSample::class.java).run {
+            addListener(listener)
+            run()
+        }
+        assertThat(listener.getPassedConfigNames()).containsExactly("bs", "bt", "bc", "bm")
+        assertThat(listener.getPassedConfigs()).hasSize(4)
+        val assertions = SoftAssertions()
+        listener.getPassedConfigs()
+            .forEach { each ->
+                assertions.assertThat(each.getAttribute(HOOK_INVOKED_ATTRIBUTE)).isNotNull
+                assertions.assertThat(each.getAttribute(HOOK_METHOD_INVOKED_ATTRIBUTE)).isNull()
+
+            }
+        assertions.assertAll()
+    }
+
+    @Test(description = "GITHUB-2257")
+    fun ensureConfigurationsCanBeRetriedViaCallBacks() {
+        val listener = TestListenerAdapter()
+        val testng = create(TestClassSample::class.java).let {
+            it.addListener(listener)
+            it.run()
+            it
+        }
+        assertThat(listener.configurationSkips).isEmpty()
+        assertThat(listener.configurationFailures).isEmpty()
+        assertThat(testng.status).isEqualTo(0)
+    }
+
+    @Test(description = "GITHUB-2266")
+    fun ensureTestsCanBeRetriedViaCallBacks() {
+        val listener = TestListenerAdapter()
+        val testng = create(test.hook.samples.issue2266.TestClassSample::class.java).let {
+            it.addListener(listener)
+            it.run()
+            it
+        }
+        assertThat(listener.failedTests).isEmpty()
+        assertThat(listener.skippedTests).isEmpty()
+        assertThat(listener.passedTests).hasSize(1)
+        assertThat(testng.status).isEqualTo(0)
+    }
+
+    @Test(description = "GITHUB-2251")
+    fun ensureTimeoutsAreHandled() {
+        val exception = create(TestCaseSample::class.java).run {
+            val l = TestListenerAdapter()
```

Comment:
Limit the number of listeners for kotlin tests: few listener should cover all needs of tests

Reply:
@juherr - That rule of thumb is usually going to be difficult to enforce because then we would end up building one listener with a whole of possibilities. The rule in general is good and its a good thing. And I am really not comfortable with doing migration and changing tests at the same time, because I may inadvertantly cause a change that kind of lets us miss a bug. 

So in these migration PRs, I would like to stick to purely migrational in nature.